x-pack/plugins/siem/public/store/timeline/epic.ts

```diff
@@ -0,0 +1,302 @@
+                    })
+                  : updateTimeline({
+                      id: get('payload.id', action),
+                      timeline: {
+                        ...savedTimeline,
+                        savedObjectId: response.timeline.savedObjectId,
+                        version: response.timeline.version,
+                        isSaving: false,
+                      },
+                    }),
+                endTimelineSaving({
+                  id: get('payload.id', action),
+                }),
+              ];
+            }),
+            startWith(startTimelineSaving({ id: get('payload.id', action) })),
+            takeUntil(
+              action$.pipe(
+                withLatestFrom(timeline$),
+                filter(([checkAction, updatedTimeline]) => {
+                  if (
+                    checkAction.type === endTimelineSaving.type &&
+                    updatedTimeline[get('payload.id', checkAction)].savedObjectId != null
+                  ) {
+                    myEpicTimelineId.setTimelineId(
+                      updatedTimeline[get('payload.id', checkAction)].savedObjectId
+                    );
+                    myEpicTimelineId.setTimelineVersion(
+                      updatedTimeline[get('payload.id', checkAction)].version
+                    );
+                    return true;
+                  }
+                  return false;
+                })
+              )
+            )
+          );
+        }
+        return empty();
+      })
+    )
+  );
+};
+
+const isNotNull = <T>(value: T | null): value is T => value !== null;
+
+const timelineInput: TimelineInput = {
+  columns: null,
+  dataProviders: null,
+  description: null,
+  kqlMode: null,
+  kqlQuery: null,
+  title: null,
+  dateRange: null,
+  sort: null,
+};
+
+const convertTimelineAsInput = (
+  timeline: TimelineModel,
+  timelineTimeRange: TimeRange
+): TimelineInput =>
+  Object.keys(timelineInput).reduce<TimelineInput>((acc, key) => {
+    if (has(key, timeline)) {
+      if (key === 'kqlQuery') {
+        return set(`${key}.filterQuery`, get(`${key}.filterQuery`, timeline), acc);
+      } else if (key === 'dateRange') {
+        return set(`${key}`, { start: timelineTimeRange.from, end: timelineTimeRange.to }, acc);
+      } else if (key === 'columns' && get(key, timeline) != null) {
+        return set(
+          key,
+          get(key, timeline).map((col: ColumnHeader) => omit(['width', '__typename'], col)),
+          acc
+        );
+      }
+      return set(key, get(key, timeline), acc);
+    }
+    return acc;
+  }, timelineInput);
+
+const omitTypename = (key: string, value: keyof TimelineModel) =>
+  key === '__typename' ? undefined : value;
+
+const omitTypenameInTimeline = (
+  oldTimeline: TimelineModel,
+  newTimeline: TimelineResult
+): TimelineModel => JSON.parse(JSON.stringify(mergeObject(oldTimeline, newTimeline)), omitTypename);
```

Comment:
I'd at least write a ticket for the backlog for writing unit tests for at least convertTimelineAsInput.

Reply:
For this stringify, I would add a try catch to guard against circular exceptions to a bit on the safer side for the stringify.